### ISSUE TITLE
使用fastllm导致推理输出不一致

### ISSUE THREAD

**Issue by lylcst** [ztxz16/fastllm#158]
`if args.use_fast:
    from fastllm_pytools import llm
    model = llm.from_hf(model, tokenizer, dtype="float16") # dtype: ["float16", "int8", "int4"]
else:
    model = model.half().cuda()
    model.eval()`

使用fastllm后用.chat和不使用fastllm输出结果有diff，do_samle设为False了，这是什么原因导致的呢？

**Comment by lylcst** [comment]
使用的chatglm2模型

**Comment by HongyuJiang** [comment]
把两边运行的随机数种子设定为一样的试试

**Comment by ztxz16** [comment]
> 应该不是随机种子的问题。我这边用 ChatGLM2-6B 模型也发现类似的问题。使用 fastllm 的模型经常在数字、标点符号、“因此”、“此外” 的后面就停了，产生一段不完整的回答，非常影响体验。而是用原版的模型，没有类似的问题，还没查出来是什么原因。

有具体输入的例子吗
用的模型是什么精度的？

**Comment by liuzhenjluccst** [comment]
> 把两边运行的随机数种子设定为一样的试试

麻烦问下具体怎么设置随机种子

**Comment by sun1092469590** [comment]
我这边是先将chatglm2-6b模型转为.flm（float16），然后用.response，没设置随机种子，结果也是看着不太好，有一些回答有复读机现象，但是原版复读机现象很少。

**Comment by HongyuJiang** [comment]
Python推理版的参考这个设置随机种子哈 https://github.com/THUDM/ChatGLM-6B/issues/277
这边C++版本的还不清楚
@liuzhenjluccst

**Comment by xxm1668** [comment]
有人解决吗？同遇到问题。。。

**Comment by ztxz16** [comment]
差异大吗？ 如果fastllm回答特别离谱可以提供一下输入
目前fastllm的中间值计算用的是float32，原版chatglm中间用的float16，暂时是不能强对齐的，不过CMMLU上的测试结果是可以提高的

**Comment by ijustloveses** [comment]
> > 应该不是随机种子的问题。我这边用 ChatGLM2-6B 模型也发现类似的问题。使用 fastllm 的模型经常在数字、标点符号、“因此”、“此外” 的后面就停了，产生一段不完整的回答，非常影响体验。而是用原版的模型，没有类似的问题，还没查出来是什么原因。
> 
> 有具体输入的例子吗 用的模型是什么精度的？

>>> from fastllm_pytools import llm
>>> from transformers import AutoTokenizer
>>> tokenizer_path = 'chatglm2-6b tokenizer 模型地址'
>>> tokenizer = AutoTokenizer.from_pretrained(tokenizer_path, trust_remote_code=True)
Explicitly passing a `revision` is encouraged when loading a model with custom code to ensure no malicious code has been contributed in a newer revision.
normalizer.cc(51) LOG(INFO) precompiled_charsmap is empty. use identity normalization.
>>> checkpoint_path = 'chatglm2-6b-fp16.flm 模型地址，float16'
>>> model = llm.model(checkpoint_path)
Load (200 / 200)
Warmup...
finish.
>>> query = "2018-2023年我国辅助生殖市场的复合增速是多少\n###与query相关的段落###\n我国作为辅助生殖的新兴市场，近年来维持较高的增速。2018年，我国辅助生殖市场达到37亿美元，与美国市场持平。2014-2018年复合 增速达到13.6%，远高于全球市场的5.1%。根据沙利文的分析，2018-2023年，我国辅助生殖市场的复合增速为14.5%，将实现加速增长，预计2023年我国辅助生殖市场有望突破300亿元。"
>>> response, _ = model.chat(tokenizer, query, do_sample=True, top_k=10, top_p=0.8, temperature=0.35, repeat_penalty=1.2)
>>> response
'我国辅助生殖市场在近年来呈现持续增长的趋势。根据沙利文的分析，2018-2023年，我国辅助生殖市场的复合增速为14.5%，这将实现加速增长。预计到20'

我在尝试一个 Retrieval Augmented Generation 场景，然后发现模型的回答会经常莫名其妙的终止，比如这个例子的生成结果。类似这样的例子蛮多的 @ztxz16

**Comment by ztxz16** [comment]
> > > 应该不是随机种子的问题。我这边用 ChatGLM2-6B 模型也发现类似的问题。使用 fastllm 的模型经常在数字、标点符号、“因此”、“此外” 的后面就停了，产生一段不完整的回答，非常影响体验。而是用原版的模型，没有类似的问题，还没查出来是什么原因。
> > 
> > 
> > 有具体输入的例子吗 用的模型是什么精度的？
> 
> '>>> from fastllm_pytools import llm '>>> from transformers import AutoTokenizer '>>> tokenizer_path = 'chatglm2-6b tokenizer 模型地址' '>>> tokenizer = AutoTokenizer.from_pretrained(tokenizer_path, trust_remote_code=True) Explicitly passing a `revision` is encouraged when loading a model with custom code to ensure no malicious code has been contributed in a newer revision. normalizer.cc(51) LOG(INFO) precompiled_charsmap is empty. use identity normalization. '>>> checkpoint_path = 'chatglm2-6b-fp16.flm 模型地址，float16' '>>> model = llm.model(checkpoint_path) Load (200 / 200) Warmup... finish. '>>> query = "2018-2023年我国辅助生殖市场的复合增速是多少\n###与query相关的段落###\n我国作为辅助生殖的新兴市场，近年来维持较高的增速。2018年，我国辅助生殖市场达到37亿美元，与美国市场持平。2014-2018年复合 增速达到13.6%，远高于全球市场的5.1%。根据沙利文的分析，2018-2023年，我国辅助生殖市场的复合增速为14.5%，将实现加速增长，预计2023年我国辅助生殖市场有望突破300亿元。" '>>> response, _ = model.chat(tokenizer, query, do_sample=True, top_k=10, top_p=0.8, temperature=0.35, repeat_penalty=1.2) '>>> response '我国辅助生殖市场在近年来呈现持续增长的趋势。根据沙利文的分析，2018-2023年，我国辅助生殖市场的复合增速为14.5%，这将实现加速增长。预计到20'
> 
> 我在尝试一个 Retrieval Augmented Generation 场景，然后发现模型的回答会经常莫名其妙的终止，比如这个例子的生成结果。类似这样的例子蛮多的 @ztxz16

这个测试的是最新版吗？ 我目前测试原版ChatGLM2-6b和fastllm模型，输出都是
"根据沙利文的分析，2018-2023年我国辅助生殖市场的复合增速为14.5%。"

**Comment by ztxz16** [comment]
之前prompt构造有点小问题，最近修复了
不过目前因为FastLLM中间层运算用的float32，原版ChatGLM6B用的是float16，因此答案确实无法强对齐
目前我这边测试CMMLU，评分比原版是上升了的，我自测好像没出现回答特别离谱的情况（如果还有可以发一下输入）

**Comment by ijustloveses** [comment]
> 之前prompt构造有点小问题，最近修复了 不过目前因为FastLLM中间层运算用的float32，原版ChatGLM6B用的是float16，因此答案确实无法强对齐 目前我这边测试CMMLU，评分比原版是上升了的，我自测好像没出现回答特别离谱的情况（如果还有可以发一下输入）

好，我后续更新了再测一下

**Comment by ijustloveses** [comment]
> > > > 应该不是随机种子的问题。我这边用 ChatGLM2-6B 模型也发现类似的问题。使用 fastllm 的模型经常在数字、标点符号、“因此”、“此外” 的后面就停了，产生一段不完整的回答，非常影响体验。而是用原版的模型，没有类似的问题，还没查出来是什么原因。
> > > 
> > > 
> > > 有具体输入的例子吗 用的模型是什么精度的？
> > 
> > 
> > '>>> from fastllm_pytools import llm '>>> from transformers import AutoTokenizer '>>> tokenizer_path = 'chatglm2-6b tokenizer 模型地址' '>>> tokenizer = AutoTokenizer.from_pretrained(tokenizer_path, trust_remote_code=True) Explicitly passing a `revision` is encouraged when loading a model with custom code to ensure no malicious code has been contributed in a newer revision. normalizer.cc(51) LOG(INFO) precompiled_charsmap is empty. use identity normalization. '>>> checkpoint_path = 'chatglm2-6b-fp16.flm 模型地址，float16' '>>> model = llm.model(checkpoint_path) Load (200 / 200) Warmup... finish. '>>> query = "2018-2023年我国辅助生殖市场的复合增速是多少\n###与query相关的段落###\n我国作为辅助生殖的新兴市场，近年来维持较高的增速。2018年，我国辅助生殖市场达到37亿美元，与美国市场持平。2014-2018年复合 增速达到13.6%，远高于全球市场的5.1%。根据沙利文的分析，2018-2023年，我国辅助生殖市场的复合增速为14.5%，将实现加速增长，预计2023年我国辅助生殖市场有望突破300亿元。" '>>> response, _ = model.chat(tokenizer, query, do_sample=True, top_k=10, top_p=0.8, temperature=0.35, repeat_penalty=1.2) '>>> response '我国辅助生殖市场在近年来呈现持续增长的趋势。根据沙利文的分析，2018-2023年，我国辅助生殖市场的复合增速为14.5%，这将实现加速增长。预计到20'
> > 我在尝试一个 Retrieval Augmented Generation 场景，然后发现模型的回答会经常莫名其妙的终止，比如这个例子的生成结果。类似这样的例子蛮多的 @ztxz16
> 
> 这个测试的是最新版吗？ 我目前测试原版ChatGLM2-6b和fastllm模型，输出都是 "根据沙利文的分析，2018-2023年我国辅助生殖市场的复合增速为14.5%。"

@ztxz16 我更新的版本，重新编译，然后这个例子输出确实 OK 了，但是我加上一些信息，又出现类似的情况，例子如下：
’>>> query = "2018-2023年我国辅助生殖市场的复合增速是多少\nInput: 文档标题: 辅助生殖行业全景图：受益人口新政的消费升级黄金赛道\n\n###与query相关的段落###\n我国作为辅助生殖的新兴市场，近年来维持较高的增速 。2018年，我国辅助生殖市场达到37亿美元，与美国市场持平。2014-2018年复合增速达到13.6%，远高于全球市场的5.1%。根据沙利文的分析，2018-2023年，我国辅助生殖市场的复合增速为14.5%，将实现加速增长，预计2023年我 国辅助生殖市场有望突破300亿元。"
‘>>> response, _ = model.chat(tokenizer, query, do_sample=True, top_k=10, top_p=0.8, temperature=0.35, repeat_penalty=1.2)
’>>> response
'根据沙利文的分析，2018-2023年我国辅助生殖市场的复合增速为14.5%，将实现加速增长，预计2023年我国辅助生殖市场有望突破3 '
